Utilities/DavixAdaptor/src/DavixFile.cc

```diff
@@ -0,0 +1,346 @@
+       << "error '" << davixErr->getErrMsg().c_str() << " and error code " << davixErr->getStatus();
+    ex.addContext("Calling DavixFile::open()");
+    throw ex;
+  }
+  if (m_fd == NULL) {
+    edm::Exception ex(edm::errors::FileOpenError);
+    ex << "Davix::open(name='" << m_name << "') failed as fd is NULL";
+    ex.addContext("Calling DavixFile::open()");
+    throw ex;
+  }
+  m_name = name;
+}
+
+IOSize DavixFile::readv(IOBuffer *into, IOSize buffers) {
+  assert(!buffers || into);
+
+  // Davix does not support 0 buffers;
+  if (buffers == 0)
+    return 0;
+
+  DavixError *davixErr;
+
+  DavIOVecInput input_vector[buffers];
+  DavIOVecOuput output_vector[buffers];
+  IOSize total = 0; // Total requested bytes
+  for (IOSize i = 0; i < buffers; ++i) {
+    input_vector[i].diov_size = into[i].size();
+    input_vector[i].diov_buffer = (char *)into[i].data();
+    total += into[i].size();
+  }
+
+  ssize_t s = davixPosix->preadVec(m_fd, input_vector, output_vector, buffers, &davixErr);
+  if (davixErr) {
+    edm::Exception ex(edm::errors::FileReadError);
+    ex << "Davix::readv(name='" << m_name << "', buffers=" << (buffers) << ") failed with error "
+       << davixErr->getErrMsg().c_str() << " and error code " << davixErr->getStatus()
+       << " and call returned " << s << " bytes";
+    ex.addContext("Calling DavixFile::readv()");
+    throw ex;
+  }
+  if (s < 0) {
+    edm::Exception ex(edm::errors::FileReadError);
+    ex << "Davix::readv(name='" << m_name << "') failed and call returned " << s;
+    ex.addContext("Calling DavixFile::readv()");
+    throw ex;
+  }
+  return total;
+}
+
+IOSize DavixFile::readv(IOPosBuffer *into, IOSize buffers) {
+  assert(!buffers || into);
+
+  // Davix does not support 0 buffers;
+  if (buffers == 0)
+    return 0;
+
+  DavixError *davixErr;
+
+  DavIOVecInput input_vector[buffers];
+  DavIOVecOuput output_vector[buffers];
+  IOSize total = 0;
+  for (IOSize i = 0; i < buffers; ++i) {
+    input_vector[i].diov_offset = into[i].offset();
+    input_vector[i].diov_size = into[i].size();
+    input_vector[i].diov_buffer = (char *)into[i].data();
+    total += into[i].size();
+  }
+  ssize_t s = davixPosix->preadVec(m_fd, input_vector, output_vector, buffers, &davixErr);
+  if (davixErr) {
+    edm::Exception ex(edm::errors::FileReadError);
+    ex << "Davix::readv(name='" << m_name << "', n=" << buffers << ") failed with error "
+       << davixErr->getErrMsg().c_str() << " and error code " << davixErr->getStatus()
+       << " and call returned " << s << " bytes";
+    ex.addContext("Calling DavixFile::readv()");
+    throw ex;
+  }
+  if (s < 0) {
+    edm::Exception ex(edm::errors::FileReadError);
+    ex << "Davix::readv(name='" << m_name << "', n=" << buffers << ") failed and call returned "
+       << s;
+    ex.addContext("Calling DavixFile::readv()");
+    throw ex;
+  } else if (s == 0)
+    return 0; // end of file
+
+  return total;
```

Comment:
What's the meaning of the return value, `s`?  Is it the number of bytes read?  Do we need a check that `s <= total`?